Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/HitSelector.cpp

```diff
@@ -0,0 +1,84 @@
+  m_inputHits.initialize(m_cfg.inputHits);
+  m_inputParticlesSelected.maybeInitialize(m_cfg.inputParticlesSelected);
+  m_outputHits.initialize(m_cfg.outputHits);
+
+  ACTS_DEBUG("selection particles " << m_cfg.inputParticlesSelected);
+  ACTS_DEBUG("selection hit x [" << m_cfg.minX << "," << m_cfg.maxX << ")");
+  ACTS_DEBUG("selection hit y [" << m_cfg.minY << "," << m_cfg.maxY << ")");
+  ACTS_DEBUG("selection hit z [" << m_cfg.minZ << "," << m_cfg.maxZ << ")");
+  ACTS_DEBUG("selection hit r [" << m_cfg.minR << "," << m_cfg.maxR << ")");
+  ACTS_DEBUG("selection hit time [" << m_cfg.minTime << "," << m_cfg.maxTime
+                                    << ")");
+  ACTS_DEBUG("selection hit energy loss [" << m_cfg.minEnergyLoss << ","
+                                           << m_cfg.maxEnergyLoss << ")");
+  ACTS_DEBUG("selection primary vertex ID [" << m_cfg.minPrimaryVertexId << ","
+                                             << m_cfg.maxPrimaryVertexId
+                                             << ")");
+}
+
+ActsExamples::ProcessCode ActsExamples::HitSelector::execute(
+    const ActsExamples::AlgorithmContext& ctx) const {
+  const SimHitContainer& hits = m_inputHits(ctx);
+  const SimParticleContainer* particlesSelected =
+      m_inputParticlesSelected.isInitialized() ? &m_inputParticlesSelected(ctx)
+                                               : nullptr;
+
+  std::vector<SimHit> unorderedHits;
+  unorderedHits.reserve(hits.size());
+
+  for (const auto& hit : hits) {
+    const double r = Acts::fastHypot(hit.position().x(), hit.position().y());
+    const std::uint64_t primaryVertexId = hit.particleId().vertexPrimary();
+
+    const bool validParticle = (particlesSelected == nullptr) ||
+                               particlesSelected->contains(hit.particleId());
+    const bool validX =
+        (m_cfg.minX <= hit.position().x()) && (hit.position().x() < m_cfg.maxX);
+    const bool validY =
+        (m_cfg.minY <= hit.position().y()) && (hit.position().y() < m_cfg.maxY);
+    const bool validZ =
+        (m_cfg.minZ <= hit.position().z()) && (hit.position().z() < m_cfg.maxZ);
+    const bool validR = (m_cfg.minR <= r) && (r < m_cfg.maxR);
+    const bool validTime =
+        (m_cfg.minTime <= hit.time()) && (hit.time() < m_cfg.maxTime);
+    const bool validEnergyLoss =
+        (m_cfg.minEnergyLoss <= hit.depositedEnergy()) &&
+        (hit.depositedEnergy() < m_cfg.maxEnergyLoss);
+    const bool validPrimaryVertexId =
+        (m_cfg.minPrimaryVertexId <= primaryVertexId) &&
+        (primaryVertexId < m_cfg.maxPrimaryVertexId);
+
+    const bool validHit = validParticle && validX && validY && validZ &&
+                          validR && validTime && validEnergyLoss &&
+                          validPrimaryVertexId;
+    if (validHit) {
+      unorderedHits.push_back(hit);
+    }
+  }
+
+  SimHitContainer selectedHits(unorderedHits.begin(), unorderedHits.end());
```

Comment:
In principle one could optimize the construction by sorting the vector and provide this `boost::container::sorted_range_t` tag (or similar), might be benefitial in for very high pile-up situations with a lot of hits...

Reply:
And I think we can reserve the space in the `SimHitContainer`, right?